### ISSUE TITLE
rpm-build: %setup and %autosetup crashes on directories with spaces

### ISSUE THREAD

**Issue by xvitaly** [rpm-software-management/rpm#1644]
I have an upstream tarball with the "Foo Bar-1.0.0" directory name.

Trying: `%autosetup -n "Foo Bar-%{version}"`, but got an error:

```
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.eAQck5
+ umask 022
+ cd /home/vitaly/rpmbuild/BUILD
+ cd /home/vitaly/rpmbuild/BUILD
+ rm -rf 'Foo  -q
'
+ /usr/bin/gzip -dc /home/vitaly/rpmbuild/SOURCES/Foo.Bar-1.0.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd 'Foo  -q
'
/var/tmp/rpm-tmp.eAQck5: line 41: cd: $'Foo  -q\n': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.eAQck5 (%prep)
```

**Comment by pmatilai** [comment]
Probably a dupe of #1445

**Comment by ffesti** [comment]
Yes, definitely a duplicate. Closing here - but working on the original issue.

**Comment by ffesti** [comment]
See #1781

**Comment by ffesti** [comment]
This needs the same kind of quoting as #1445 but at other places. %patches is not actually used by %autosetup.

**Comment by pmatilai** [comment]
Yup. I bet there are plenty of places needing that kind of quoting, only most of the world is sane enough not to use spaces and special characters like quotes in file names so nobody ever notices.

**Comment by ffesti** [comment]
Ok, more thorough investigation turns up a few more problems than expected:

- Macro parameters are split by ` splitQuoted()` https://github.com/rpm-software-management/rpm/blob/master/rpmio/macro.c#L930
- It uses Unit Separator (ASCII 0x1f) as separator and ignores quotes
- Non UTF-8 characters are no longer legal in spec files...
- Even if they where legal one would still need normal quotes inside as the Unit Separators would be stripped during parsing and to allow the `%setup` macro inside to understand it is in one piece.
- While normal macros use `getopt` the build-in macros use `popt` for parameter handling which actually does handle quotes.

**Comment by ffesti** [comment]
OK, to be more clear with this: We currently don't have any (working) means to have spaces in macro parameters (with the exception of built-in macros which are different in all kind of wonderful ways).

I wonder if we actually need two ways for doing that. One that is stripped during parsing (as the old Unit Separator thing was) and one that stays in place. This would allow passing the data through multiple layers of macro expansion without getting into quoting hell. This should be made compatible with shell parameters so it does not need to get stripped off after the expansion.

**Comment by ffesti** [comment]
Skipping over `"\ "` and `"\^t"` in `splitQuoted()` makes `%autosetup -n auto\ foo` work. Still need to figure out what this all breaks. It is also not compatible with `shescape` which uses single quotes.

**Comment by pmatilai** [comment]
> * It uses Unit Separator (ASCII 0x1f) as separator and ignores quotes
> 
> * Non UTF-8 characters are no longer legal in spec files...

The unit separators do not end up in the spec, or anywhere else in the package. They only ever exist inside the macro engine.

**Comment by pmatilai** [comment]
But okay, indeed quote only really works for macro *arguments*. Somehow, option parameters were never even considered.

**Comment by pmatilai** [comment]
BTW, there's this 2001 vintage comment in macros.c :sweat_smile:  https://github.com/rpm-software-management/rpm/blob/470498bd5a51f8d98ae8e721beea58ef81c19a51/rpmio/macro.c#L830
